### PR TITLE
Issue 128 hyperparam

### PR DESCRIPTION
**Pull request by cchoquette**
#### Description
Enable manual hyperparameter tuning via serialized objects. 
Enable future use of automated tuning.
Includes ParamSpec, which holds 1 search space (a run of optimization of the different choices of that search space).
SearchOptimizer which holds all search spaces and the algorithms used to search in them.

### CODE REVIEW

File: foreshadow/wrapper.py [hunk 1]
```diff
@@ -129,57 +128,27 @@ def get_params(self, deep=True):
...
-                        self_params.update(
-                            (key + "__" + k, val) for k, val in deep_items
-                        )
-                    self_params[key] = value
-
-            return self_params
+            params = super().get_params(deep=deep)
+            return params
+
+        def set_params(self, **params):
```
Comment (JingJZ160) [inline foreshadow/wrapper.py]: Do we need this? It's just using super.

File: foreshadow/concrete/internals/fancyimpute.py [hunk 2]
```diff
@@ -43,7 +53,7 @@ def get_params(self, deep=True):
             dict: Parameter names mapped to their values.
 
         """
-        return {"method": self.method, "impute_kwargs": self.impute_kwargs}
+        return super().get_params(deep=deep)
 
     def set_params(self, **params):
         """Set the parameters of this estimator.
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/concrete/internals/fancyimpute.py]: done.

File: foreshadow/estimators/auto.py [hunk 3]
```diff
@@ -315,6 +315,31 @@ def score(self, X, y, sample_weight=None):
         y = check_df(y)
         return self.estimator.score(X, y)
 
+    def get_params(self, deep=True):
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/estimators/auto.py]: done.

File: foreshadow/estimators/auto.py [hunk 4]
```diff
@@ -315,6 +315,31 @@ def score(self, X, y, sample_weight=None):
...
+            deep: True to recursively call get_params, False to not.
+
+        Returns:
+            params for this object.
+
+        """
+        params = super().get_params(deep=deep)
+        return params
+
+    def set_params(self, **params):
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/estimators/auto.py]: done

File: foreshadow/foreshadow.py [hunk 5]
```diff
@@ -38,16 +40,25 @@ class Foreshadow(BaseEstimator):
...
         self.estimator = estimator
         self.optimizer = optimizer
+        self.optimizer_kwargs = (
+            {} if optimizer_kwargs is None else optimizer_kwargs
+        )
         self.pipeline = None
         self.data_columns = None
 
         if isinstance(self.estimator, AutoEstimator) and optimizer is not None:
+            # TODO implement V2 architecture here.
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/foreshadow.py]: done.

File: foreshadow/foreshadow.py [hunk 6]
```diff
@@ -160,7 +173,17 @@ def optimizer(self, o):
         if o is None or (inspect.isclass(o) and issubclass(o, BaseSearchCV)):
             self._optimizer = o
         else:
-            raise ValueError("Invalid value passed as optimizer")
+            raise ValueError("Invalid optimizer: '{}' passed.".format(o))
+
+    def _reset(self):
+        try:
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/foreshadow.py]: done.

File: foreshadow/optimizers/param_distribution.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+import hyperopt.hp as hp
+
+import foreshadow.serializers as ser
+
+from .tuner import _replace_list, get
+
+
+"""
+2. cases:
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/optimizers/param_distribution.py]: oh thank you! removed.

File: foreshadow/optimizers/param_distribution.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            raise ValueError(
+                "Either all kwargs are None or all are set. To "
+                "use automatic param determination, pass all "
+                "kwargs. Otherwise, manual setting can be "
+                "accomplished using set_params."
+            )
+        self._param_set = False
+        self.param_distributions = []
+
+        if not (fs_pipeline is None) and (X_df is None) and (y_df) is None:
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/optimizers/param_distribution.py]: No this is the automatic pipelining.

File: foreshadow/optimizers/random_search.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            iterable of unique states.
+
+        """
+        # check if all distributions are given as lists
+        # in this case we want to sample without replacement
+        rng = check_random_state(self.random_state)
+        prev_samples = []
+        max_tries = self.max_tries if self.max_tries is not None else 1
+        for _ in six.moves.range(self.n_iter):
+            # import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/optimizers/random_search.py]: done.

File: foreshadow/optimizers/random_search.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        Returns:
+            iterable of unique states.
+
+        """
+        # check if all distributions are given as lists
+        # in this case we want to sample without replacement
+        rng = check_random_state(self.random_state)
+        prev_samples = []
+        max_tries = self.max_tries if self.max_tries is not None else 1
+        for _ in six.moves.range(self.n_iter):
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/optimizers/random_search.py]: done.

File: foreshadow/optimizers/tuner.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        to_replace = {}
+        for key, v in obj.items():
+            to_replace[key] = _replace_list(key, v, replace_with)
+        obj.update(to_replace)
+        return obj
+    else:  # no nesting and no need to replace.
+        return obj
+
+
+def get(optimizer, **optimizer_kwargs):
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/optimizers/tuner.py]: I had it there initially but it's optimizers.get() so I think its explicit enough.

File: foreshadow/serializers.py [hunk 12]
```diff
@@ -269,7 +269,8 @@ def dict_serialize(self, deep=True):
 
         """
         return _make_serializable(
-            self.get_params(deep), serialize_args=self.serialize_params
+            self.get_params(deep),
+            # serialize_args=self.serialize_params
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/serializers.py]: done.

File: foreshadow/serializers.py [hunk 13]
```diff
@@ -448,6 +447,47 @@ def dict_serialize(self, deep=False):
         return super().dict_serialize(deep=deep)
 
 
+class ParamSpecSerializerMixin(ConcreteSerializerMixin):
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/serializers.py]: removed.

File: foreshadow/smart/smart.py [hunk 14]
```diff
@@ -150,20 +142,6 @@ def get_params(self, deep=True):
...
-            )
-        params.update(transformer_params)
-        params = {
-            key: val
-            for key, val in params.items()
-            if key.find("transformer__") == -1
-        }
         return params
 
     def set_params(self, **params):
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/smart/smart.py]: removed.

File: foreshadow/steps/preparerstep.py [hunk 15]
```diff
@@ -454,8 +456,7 @@ def fit(self, X, *args, **kwargs):
         """
         # TODO make fit remove a step if nothing is done, rather than a
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/steps/preparerstep.py]: yes but that is wel on the radar with serialization.

File: foreshadow/tests/test_estimators/test_meta.py [hunk 16]
```diff
@@ -72,3 +76,23 @@ def test_metaestimator_score():
...
+
+    """
+    from foreshadow.estimators.meta import MetaEstimator
+    from sklearn.preprocessing import StandardScaler
+    from sklearn.linear_model import LinearRegression
+
+    me = MetaEstimator(LinearRegression(), StandardScaler())
+    params = me.get_params(deep=deep)
+
+    desired_keys = ["estimator", "preprocessor"]
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/tests/test_estimators/test_meta.py]: No it isn't, just a preprocessing step. Terminology is just confusing here and should be looked at.

File: foreshadow/tests/test_optimizers/test_random_search.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+    Args:
+        estimator: fixture estimator
+        simple_distribution: fixture distribution to parameter optimize on.
+        iris_data: fixture dataset to use.
+
+    """
+    from foreshadow.optimizers import RandomSearchCV
+
+    estimator, counter = estimator
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/tests/test_optimizers/test_random_search.py]: fixed.

File: foreshadow/tests/test_optimizers/test_random_search.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    rs = RandomSearchCV(estimator=estimator, param_distributions=dist)
+    rs.fit(X, y)
+    unique_samples = set()
+    for sample in counter:
+        v = ""
+        for val in sample.values():
+            v += str(val)
+        unique_samples.add(v)
+
+    assert len(unique_samples) == 4
```
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/tests/test_optimizers/test_random_search.py]: issue to fix this.

**Pull request merged**